drivers/adc/adc_stm32.c

```diff
@@ -1174,6 +1176,58 @@ static const struct adc_driver_api api_stm32_driver_api = {
 	.ref_internal = STM32_ADC_VREF_MV, /* VREF is usually connected to VDD */
 };
 
+/*
+ * Prefix default when not configured by the DTS
+ * depending on the serie (stm32F1 has none)
+ */
+#if defined(CONFIG_SOC_SERIES_STM32L1X) || \
+	defined(CONFIG_SOC_SERIES_STM32U5X)
+#define prescaler_default	LL_ADC_CLOCK_ASYNC_DIV
+#elif !defined(CONFIG_SOC_SERIES_STM32F1X)
+#define prescaler_default	LL_ADC_CLOCK_SYNC_PCLK_DIV
+#endif
+
+/* st_prescaler property requires 2 elements : clock ASYNC/SYNC and DIV */
+#define STM32_ADC_CLOCK(x, prop)	DT_INST_PROP_BY_IDX(x, prop, 0)
+#define STM32_ADC_DIV(x, prop)		DT_INST_PROP_BY_IDX(x, prop, 1)
+
+/*
+ * Macro to check the property is valid with two elements:
+ * 1st one is SYNC or ASYNC (not all series have both)
+ * 2nd one for the divider (not all series have all values)
+ */
+#define ADC_STM32_HAS_PROP(x, prop)					\
+	IF_ENABLED(DT_INST_NODE_HAS_PROP(x, prop),			\
+		(COND_CODE_1(						\
+			IS_EQ(DT_INST_PROP_LEN(x, prop), 2),		\
+			(BUILD_ASSERT((					\
+				(STM32_ADC_CLOCK(x, prop) == SYNC) ||	\
+				(STM32_ADC_CLOCK(x, prop) == ASYNC)),	\
+				"Invalid ADC "#prop" property")),	\
+			(BUILD_ASSERT((DT_INST_PROP_LEN(x, prop) == 2),	\
+				"Invalid ADC "#prop" property length")))))
+
+/* Macro to set the prefix depending on the 1st element: check if it is SYNC or ASYNC */
+#define STM32_ADC_PREFIX(x, prop)				\
+	COND_CODE_1(IS_EQ(STM32_ADC_CLOCK(x, prop), SYNC),	\
+		LL_ADC_CLOCK_SYNC_PCLK_DIV,			\
+		LL_ADC_CLOCK_ASYNC_DIV)
+
+/*
+ * Concat prefix (1st element) and DIV value (2nd element) of st,adc-prescaler
+ * not available for stm32F1 serie: set the config->clock_prescaler to 0 instead
+ * If the property is not defined, the default prescaler with DIV 4 is used.
+ */
+#if defined(CONFIG_SOC_SERIES_STM32F1X)
+#define DT_STM32_PRESC_PROP(x, prop) 0
+#else
+#define DT_STM32_PRESC_PROP(x, prop)					\
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(x, prop),			\
+	_CONCAT(STM32_ADC_PREFIX(x, prop), STM32_ADC_DIV(x, prop)),	\
+	_CONCAT(prescaler_default, 4))
```

Comment:
All these macros are specific to this prescaler property.
I think we should remove the 'prop' argument, since no other prop could be used as argument.

Reply:
changing macro name to `DT_STM32_PRESC`